File: Trivia/Trivia/client_gui/OTPCryptoAlgorithm.cs
using System;
using System.IO;
using System.Security.Cryptography;
using System.Text;
using Newtonsoft.Json;

namespace client_gui.Crypto
{
    /// <summary>
    /// One-Time Pad encryption implementation
    /// </summary>
    public class OTPCryptoAlgorithm : CryptoAlgorithm
    {
        private string _userId;
        private byte[] _key;
        private int _position;
        private static readonly string KeyDirectory = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
            "TriviaGame", "Keys");

        /// <summary>
        /// Creates a new OTP crypto algorithm instance for a specific user
        /// </summary>
        /// <param name="userId">The user ID to use for key storage</param>
        public OTPCryptoAlgorithm(string userId)
        {
            _userId = userId;

            // Ensure key directory exists
            if (!Directory.Exists(KeyDirectory))
            {
                Directory.CreateDirectory(KeyDirectory);
            }

            if (!LoadKeyFromStorage())
            {
                GenerateNewKey();
            }
        }

        /// <summary>
        /// Encrypts a message using the OTP
        /// </summary>
        public override string Encrypt(string message)
        {
            if (_key == null || _key.Length == 0)
            {
                Console.WriteLine("OTP key is empty! Encryption failed.");
                return message;
            }

            // Check if we have enough key material left
            if (_position + message.Length > _key.Length)
            {
                GenerateNewKey(Math.Max(1024, message.Length * 2));
            }

            // Perform XOR operation with key starting at current position
            byte[] messageBytes = Encoding.UTF8.GetBytes(message);
            byte[] result = new byte[messageBytes.Length];

            for (int i = 0; i < messageBytes.Length; i++)
            {
                // Apply XOR with the key at the current position
                int keyIndex = (_position + i) % _key.Length;
                result[i] = (byte)(messageBytes[i] ^ _key[keyIndex]);
            }

            // Update position in the key
            _position += messageBytes.Length;
            SaveKeyToStorage();

            // Return Base64 encoded string
            return Convert.ToBase64String(result);
        }

        /// <summary>
        /// Decrypts a message using the OTP
        /// </summary>
        public override string Decrypt(string message)
        {
            if (_key == null || _key.Length == 0)
            {
                Console.WriteLine("OTP key is empty! Decryption failed.");
                return message;
            }

            try
            {
                // Convert Base64 string back to bytes
                byte[] encryptedBytes = Convert.FromBase64String(message);
                byte[] result = new byte[encryptedBytes.Length];

                // For OTP, encrypt and decrypt operations are identical (XOR is its own inverse)
                for (int i = 0; i < encryptedBytes.Length; i++)
                {
                    int keyIndex = (_position + i) % _key.Length;
                    result[i] = (byte)(encryptedBytes[i] ^ _key[keyIndex]);
                }

                // Update position in the key
                _position += encryptedBytes.Length;
                SaveKeyToStorage();

                return Encoding.UTF8.GetString(result);
            }
            catch (FormatException ex)
            {
                Console.WriteLine($"Decryption error: {ex.Message}");
                return message; // Return original message if decryption fails
            }
        }

        /// <summary>
        /// Generates a new random key
        /// </summary>
        /// <param name="keyLength">The length of the key to generate</param>
        public bool GenerateNewKey(int keyLength = 1024)
        {
            // Generate a random key of specified length
            _key = new byte[keyLength];
            using (var rng = new RNGCryptoServiceProvider())
            {
                rng.GetBytes(_key);
            }

            // Reset position to 0
            _position = 0;

            // Save the new key to storage
            return SaveKeyToStorage();
        }

        /// <summary>
        /// Gets the current position in the key
        /// </summary>
        public int GetKeyPosition()
        {
            return _position;
        }

        /// <summary>
        /// Loads the key from storage
        /// </summary>
        private bool LoadKeyFromStorage()
        {
            try
            {
                string filePath = Path.Combine(KeyDirectory, $"{_userId}.key");
                if (!File.Exists(filePath))
                {
                    return false;
                }

                string json = File.ReadAllText(filePath);
                var keyData = JsonConvert.DeserializeObject<KeyData>(json);

                _key = Convert.FromBase64String(keyData.Key);
                _position = keyData.Position;

                Console.WriteLine($"Loaded OTP key for user {_userId}, position: {_position}");
                return true;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error loading OTP key: {ex.Message}");
                return false;
            }
        }

        /// <summary>
        /// Saves the key to storage
        /// </summary>
        private bool SaveKeyToStorage()
        {
            try
            {
                string filePath = Path.Combine(KeyDirectory, $"{_userId}.key");

                var keyData = new KeyData
                {
                    Key = Convert.ToBase64String(_key),
                    Position = _position
                };

                string json = JsonConvert.SerializeObject(keyData);
                File.WriteAllText(filePath, json);

                return true;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error saving OTP key: {ex.Message}");
                return false;
            }
        }

        /// <summary>
        /// Key data storage model
        /// </summary>
        private class KeyData
        {
            public string Key { get; set; }
            public int Position { get; set; }
        }
    }
}

File: Trivia/Trivia/Trivia/OTPCryptoAlgorithm.cpp
// OTPCryptoAlgorithm.cpp
#include "OTPCryptoAlgorithm.h"
#include <iostream>
#include <random>
#include <algorithm>

OTPCryptoAlgorithm::OTPCryptoAlgorithm(IDatabase* database, const std::string& userId)
    : m_database(database), m_userId(userId), m_position(0)
{
    if (!loadKeyFromDatabase()) {
        // If key doesn't exist, generate a new one
        generateNewKey();
    }
}

std::string OTPCryptoAlgorithm::encrypt(const std::string& message)
{
    if (m_key.empty()) {
        std::cerr << "OTP key is empty! Encryption failed." << std::endl;
        return message; // Return the original message if encryption fails
    }

    // Check if we have enough key material left
    if (m_position + message.length() > m_key.length()) {
        // Generate a new key if we're running out
        generateNewKey(std::max(static_cast<size_t>(1024), message.length() * 2));
    }

    // Perform XOR operation with key starting at current position
    std::string encrypted = performXOR(message, m_position);

    // Update position in the key
    m_position += message.length();
    saveKeyToDatabase(); // Save the updated position

    return encrypted;
}

std::string OTPCryptoAlgorithm::decrypt(const std::string& message)
{
    if (m_key.empty()) {
        std::cerr << "OTP key is empty! Decryption failed." << std::endl;
        return message; // Return the original message if decryption fails
    }

    // For OTP, encrypt and decrypt operations are identical (XOR is its own inverse)
    // The key position should already be at the right place for the recipient
    std::string decrypted = performXOR(message, m_position);

    // Update position in the key
    m_position += message.length();
    saveKeyToDatabase(); // Save the updated position

    return decrypted;
}

bool OTPCryptoAlgorithm::generateNewKey(size_t keyLength)
{
    // Generate a random key of the specified length
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(0, 255);

    m_key.clear();
    m_key.reserve(keyLength);

    for (size_t i = 0; i < keyLength; ++i) {
        m_key.push_back(static_cast<char>(dis(gen)));
    }

    // Reset position to 0
    m_position = 0;

    // Save the new key to the database
    return saveKeyToDatabase();
}

size_t OTPCryptoAlgorithm::getKeyPosition() const
{
    return m_position;
}

bool OTPCryptoAlgorithm::loadKeyFromDatabase()
{
    try {
        // Try to load existing key
        std::string key;
        size_t position;

        if (m_database->getUserCryptoKey(m_userId, key, position)) {
            m_key = key;
            m_position = position;
            return true;
        }

        return false;
    }
    catch (const std::exception& e) {
        std::cerr << "Error loading key from database: " << e.what() << std::endl;
        return false;
    }
}

bool OTPCryptoAlgorithm::saveKeyToDatabase()
{
    try {
        // Save the key and current position
        return m_database->saveUserCryptoKey(m_userId, m_key, m_position);
    }
    catch (const std::exception& e) {
        std::cerr << "Error saving key to database: " << e.what() << std::endl;
        return false;
    }
}

std::string OTPCryptoAlgorithm::performXOR(const std::string& message, size_t startPosition)
{
    std::string result;
    result.reserve(message.length());

    for (size_t i = 0; i < message.length(); ++i) {
        // XOR each character with corresponding key byte
        result.push_back(message[i] ^ m_key[startPosition + i]);
    }

    return result;
}